src/lua/serializer.c

```diff
@@ -559,7 +564,43 @@ luaL_tofield(struct lua_State *L, struct luaL_serializer *cfg, int index,
 			} else if (ctypeid == CTID_UUID) {
 				field->ext_type = MP_UUID;
 				field->uuidval = (struct tt_uuid *) cdata;
-			} else if (ctypeid == CTID_CONST_STRUCT_ERROR_REF) {
+			} else if (ctypeid == CTID_CONST_STRUCT_ERROR_REF &&
+				   (!cfg->encode_error_as_ext ||
+				    !box_error_serialize_verbose)) {
```

Comment:
Actually, right now, the serialization metamethod never gets called, because the serializers always take this branch, as I described in the comments.

The ticket contained sample output to the console with the serialization metamethod apparently called, so I assumed we want to make the serializers use the serialization metamethod now.

If we leave it as is, the patch will be mostly useless, because one would have to manually call the serialize metamethod, or use custom serializers.

/cc @locker

Reply:
I see, changing only `__serialize` is not enough.

I discussed this with a @locker. We don't need to change `msgpack` or `msgpackffi` encoders. We can also leave `json` for now. We only need to change `yaml` encode behavior.

Changing `luaL_tofield` behavior this way does not looks right. We can change `yaml` encoder similar to how it is done in `msgpack` encoder for handling `encode_error_as_ext` using `luaL_convertfield`.

Yet a more proper solution would be to make `luaL_checkfield` call `__serialize` for `MP_EXT` object like it does now for `udata` I guess. Then we won't touch `msgpack` because it uses `luaL_tofield`. This way we can drop special treatment of `MP_EXT` in `json` and `lua` encoders and move it to `cdata` `__serialize` method.

Reply:
> Yet a more proper solution would be to make `luaL_checkfield` call `__serialize` for `MP_EXT` object like it does now for `udata` I guess.

Other extension types (decimal, uuid, etc.) only have the string conversion metamethod. Technically, the second approach would still work (I don't know who would want to disable the `encode_use_string` option):
https://github.com/tarantool/tarantool/blob/d9e0b58aac518b4ac2d6df57c450b9ccacb40933/src/lua/serializer.c#L616-L636

> Then we won't touch msgpack because it uses `luaL_tofield`.

No, it also uses `luaL_tofield` when `encode_error_as_ext` is false:
https://github.com/tarantool/tarantool/blob/d9e0b58aac518b4ac2d6df57c450b9ccacb40933/src/lua/msgpack.c#L343-L347

https://github.com/tarantool/tarantool/blob/d9e0b58aac518b4ac2d6df57c450b9ccacb40933/src/lua/msgpack.c#L373-L376